libethash-cuda/CUDAMiner.cpp

```diff
@@ -513,9 +513,14 @@ void CUDAMiner::search(
 				}
 			}
 			addHashCount(batch_size);
-			if (m_abort || shouldStop())
+			bool t = true;
+			if (m_abort.compare_exchange_weak(t, false))
```

Comment:
I think we just need single `m_abort.store(false)` here.

Reply:
Yes. Simpler.. better

Reply:
Actually, this one needs to stay the way it is. We don't want to break out if m_abort is not set. compare_exchange returns true only if the exchange was successful. if m_abort is false here, the exchange will fail, which is what we want.

Reply:
Ok, that make sense, but you need strong exchange. The weak one can also fail if the abort flag is true.

Reply:
We want it to fail! in the case where m-abort is false, the exchange will fail, it will return false, and we'll keep looping as we should. We only want the exchange to succeed if m_abort is true.

Reply:
That's not what I mean.

> We only want the exchange to succeed if m_abort is true.

Ok, that's right. But `compare_exchange_weak` can sometimes (probably not the case on x86 anyway) fail randomly. So even "if m_abort is true" it will fail to exchange the values. This is just an optimization for spinlocks where you loop anyway. In our case, yes, we loop, but I'm not sure we should schedule wrong kernel execution just for a small (if any) performance gain. I suggest using `compare_exchange_strong` -- the same semantic, but will never randomly fail to exchange the value if the expected value is valid.

Reply:
Yes it can. I think that's the whole idea! The strong compare is just a looping version of weak one that retries till it succeeds. Fail, in this case, does not mean it failed to maintain atomicity, it just means that the compare did not match so the exchange didn't take place.

Reply:
Isn't this a problem?

> The weak forms (1-2) of the functions are allowed to fail spuriously, that is, act as if *this != expected even if they are equal. When a compare-and-exchange is in a loop, the weak version will yield better performance on some platforms.

Reply:
That is only true for architecture like MIPS. As you point out it won't happen on x86. But if it did, we can tolerate spurious failures here, this loop will retry in microseconds. Will the strong exchange return false on a compare failure? We can't afford to get stuck here!

Reply:
I'll do a quick test...

Reply:
Strong is ok. I'll update.

Reply:
Just to make it clear: compare_exchange_strong does not loop waiting for expected value to match (neither compare_exchange_weak does).

Reply:
@chfast Yes. I misunderstood the semantics for weak vs. strong.